### ISSUE TITLE
IsAuthorized DB design (Part 2)

### ISSUE THREAD

**Issue by adelefelicia** [indykite/neo4j-graph-tool-core#205]
**Created by: @pepe-zamarripa Jun 21 2022 at 15:27 CET**
Last updated: Jul 19 2022 at 15:06 CET
ClickUp ID: IK-3429

Given
A subject is asking for access to a resource

When
IsAuthorized(subject, resource, access) is called

Then
DB should run a cypher, which returns with an appropriate response. 

Oslo Discussion Points:
SVO Triples:
We started to discuss the problem from the SVO classic triple (Subject-Verb-Object) perspective; each of these three entities might be itself a triple, group or subgraph. A Policy ties these up together.
Aggregation/Grouping:
A consequence of this complexity is aggregation (or grouping) of identities into roles, actions into permissions, resources into assets.  We need a way to capture those, when querying the graph, without knowing their definition a-priori (customer-defined). 
KGraph Schema:
The KG schema (Schema) has been said to solve policy semantic constraints, but that is not totally true as argued later with Blueprints.

The 3 Models:
Permission node and Action link
Just Action link
No Permission-Action policy explicitly stored in db (Permission-Less Model) (based on customer schema)
This last one is our best candidate, because it decouples policies from the KG data the policies apply to: 
the schema drives part of the authorization semantics ➝ The authorization problem becomes a reachability problem at schema level
there must be no modification to the graph itself (due to conflicts) ➝ consider projections (GDS) in the future?
we do not implement negative policies, so just positive ones (modelling what User is allowed, not denied to) ➝ for example the subject is a kid in Rema, the resources the 18+ items in the supermarket. Instead of having a policy that says "DT with age<18 cannot access", we would have "DT with age>=18 can access".
cypher generation from graph schema+policy
natural customer interaction with data model, minimal concept translation (visual, nl), to be used for generating the correct cyphers ➝ this maybe could be solved with policy flows (see below)
OPA and Zanzibar (part1):
Zanzibar: 
- specific indexes for nested triples
- user-sets, 
- zookies, 
- distributed and highly available
OPA: 
- set operators, 
- composition, 
- ... 
What we take from these:
being highly available and optimized for performance by design
specific indexes and abstraction of structures, be it in the KG or in the policy data management.
Where is the policy defined?
Partly in the KG schema ➝ conflicts and sync bw policies and schema: schema cannot be changed for policies
partly in the policy data (for now we consider the params listed below as predicates for the policy language). 
The returned resources are thus the result of a mix of a graph reachability problem and resource subgraph search.
Problems:
Using the KG schema as it is for authorization ground truth will inevitably lead to conflicts: it embeds business data semantics but lacks policy data semantics. We cannot enforce the two to be coincident in defining policies. Thus this is not just a problem of reachability in the KG.
Need of properly separating two overlapping resource sets being used by two policies. There are situations in which the customer would not define this separation in the KG simply because it does not convey the meaning from business side, it would just in the permission sense. But there would be an abstraction of it that would have also authorization value.
The policy needs an abstraction as well; to not fall back to PBAC, policies should be flexible enough.

Possible Implementation
We try to get to the solution of this by building it up from the other end: the object or the resource set. This has also been the approach taken for Zanzibar, but in our case authorization is at another level and cannot be solved with their tool alone.
There are two sub-problems: resource set search and graph reachability dt➝ resources
resource set search: find the set of resources involved in the policy
graph reachability: we can use directionality from DT to the boundary of the resource set
As said, the KG schema as-is cannot fit the policy semantics. An edited version of it is needed that would give more flexibility to the policy definition. Let's call it a Blueprint, such that we have:
Schema: KG schema, is the business data schema
Blueprint: modified version of the KG schema to fit policy semantics
This does not mean managing many blueprints. We can apply some reasonable generic edits to fit all policies, but then we need to know how to define these policies.
We also have to be able to handle and possibly resolve the most scenarios without the customer intervention, plus to prove the maintenance is at least as precise as having a KG model that has permission/policy data embedded in the graph.

There are three possible macro changes during a business operation:
schema change 
policy change
resource set change
Since these will interfere with each other, let's consider the resource set change as the driver of the changes.
3. RS change: 
affect schema change (nodes/rels 
- affect policy change ➝ 
- no policy change ➝
no schema change (nodes/rels added to RS, already existing in the schema) [rel directionality matters for RS too?]toedit
- with policy change ➝  another label used to define subgroup toedit
- without policy change ➝ for example another label is used to define a subgroup of an already existing label, this is recognized and the policy is not touched?don't think we need to automatically do this.

Policy Definition
Subject: set of DTs (we now assume one DT to validate access request)
Verb: It is now assumed a unique verb, HAS_ACCESS, not in the KG nor the Policy data; we can assume it will be in the latter. 
Policy: wires S,V and O up together, relegating the Verb function to indicate the properties of the action, while giving policies the structured key to link S and O in the KG. In FXC there are specific relationships (CEO, Owner) defining role/function of DT over a starting resource (Company, Shares).
Object: set of Resources, with a dynamic configuration.
Blueprint: undirected to abstract the reachability problem from KG semantics; handling multi/sublabels not captured by the schema; optionally identify (and extend with depth/sequence?) self-loops and/or loops in general, plus sequences of nodes/rels/subgraphs TOEDIT is this valuable for policy semantics actually?

FXC KG:
FXC mock small graph with 4 DTS, of which 2 are CEOs of Companies and all are Owners of Shares. 1 DT is double Owner.

The KG schema (or Schema), in this case equivalent to the Policy schema(Blueprint). For the latter, the graph should be considered undirected for the resource subgraph (Shares and Company with interconnections):

We extend slightly the schema to include Companies owning other Companies:

new schema:

Schema Change:
added self-loop in Company ➝ potentially infinite chains of Companies, which policy needs a depth range and which not?
DT-Resources path
Resource Set Change

Policies:
CEO of X access [X, stocks/companies of X].

Owner of stock of X access [X, owned stock]

Simple test cypher:
We just solve the reachability DT-(resources) and give back the connected ones. Policies are here cyphers, can see an example with FXC use case.
If this all boiled down to a reachability problem (it does not), then the cypher would be:
:params 
{
    dt_id:'002',
    resources:['ext_01','ext_02','ext_03','ext_04', 'ext_21', 'ext_31', 'ext_41', 'ext_d11', 'ext_d21'],   
    action:'CEO',
    start_node_label:'Company',
    actions:['CEO']
}

MATCH (dt:DigitalTwin{id: $ dt_id})-[r]->(sn)
WHERE TYPE(r)= $ action AND  $ start_node_label IN LABELS(sn)
WITH sn
UNWIND  $ resources AS resource
MATCH(sn)-[*]-(res)
WHERE res.externalId=resource 
WITH DISTINCT res.externalId AS resource_id
RETURN resource_id
the first three lines define the cypher "role" in the policy (DT is CEO of Company), the resources returned are ALL those reachable by the DT (via the Company). Those are all the connected nodes to the starting one, if the two starting nodes of two different policies are connected, then those two policies will always found the same set of nodes, even if their intent is pretty different (CEO vs Owner).
Cypher+APOC test cypher:
So to solve the policies above we might add the use of apoc.subgraph:
CEO of X access [X, stocks/companies of X].
:params {
   init_cypher:{
      dt_id:'002',
      action:'CEO',
      start_node_label:'Company'
   },
   subgraph:{
      relationshipFilter: 'OF',
      labelFilter:'+Shares|+Company',
      bfs:true, 
      filterStartNode:true, 
   },
   resources:['ext_01','ext_02','ext_03','ext_04', 'ext_21', 'ext_31', 'ext_41', 'ext_d11', 'ext_d21']
}

These are the correct entities to show. W has id='ext_01'.

Owner of stock of X access [company X, owned stock]
:params {
   init_cypher:{
      dt_id:'002',
      action:'OWNER',
      start_node_label:'Shares'
   },
   subgraph:{
      relationshipFilter: 'OF',
      labelFilter:'+Company',
      bfs:true, 
      filterStartNode:true, 
   },
   resources:['ext_01','ext_02','ext_03','ext_04', 'ext_21', 'ext_31', 'ext_41', 'ext_d11', 'ext_d21']
}

These are the correct entities to show. ids here below.
[ext_d26, ext_06, ext_d21, ext_01]

If we then also compare the initial resource id list given and the resulting from the policy, we obtain this query, in which we return the intersection of the two lists:
MATCH (dt:DigitalTwin{id: $ init_cypher.dt_id})-[r]->(sn)
WHERE TYPE(r)= $ init_cypher.action AND  $ init_cypher.start_node_label IN LABELS(sn)
WITH sn
CALL apoc.path.subgraphAll(
    sn, 
    {
      relationshipFilter: $ subgraph.relationshipFilter,
      labelFilter: $ subgraph.labelFilter,
      bfs:true, 
      filterStartNode:false
    })
YIELD nodes,relationships
WITH nodes,relationships
RETURN apoc.coll.intersection([n IN nodes WHERE n.externalId IS NOT NULL],  $ resources) AS allowed

----------------------------------------------------------------------------------------------------------
//NOTES
//listcompr RETURN [x IN range(0,10) WHERE x % 2 = 0 ] AS result
// pattcompr RETURN [(a)-->(b:Movie) WHERE b.title CONTAINS 'Matrix' | b.released] AS years
// WITH [(a)-->(b:Movie) | b.released] AS years
//UNWIND years AS year
//WITH year ORDER BY year
//RETURN COLLECT(year) AS sorted_years
//MATCH(sn)-[*]-(res)
//WHERE res.externalId=resource AND  $ res_label IN LABELS(res)
//WITH DISTINCT res.externalId AS resource_id
//RETURN resource_id
Next with this approach: 
test RS and Schema changes ➝ the first ones we have to handle and report, the second to halt and report
different actions and resource node properties
by MB, represent:
define action based on many relationships?
eg:
(:Subject)-[one or many]-(one or many)-[action]-(:Resource)
(:Subject)-[action1]-(one or many)-[action2]-(:Resource)
(:Subject)-(one or many nodes with many relationships action)-(:Resource)

Big Filter Cypher:
If the above does not work, a big cypher with subqueries handling different filters (tools) may be used to define the paths between DT and Resources. Have a draft containing pieces of it below (Query), but the logic would have been hard to implement without issues, plus constrained to what a single cypher could do.
One built up from a sequence of smaller ones, connected by an operator, would give a way higher configurability, especially if could be visual. While this last idea is the basis for Policy Flows, let's look at how the big filter cypher would look like:
Query Params: toedit
Subgraph parameters (a bunch, see the cypher)
Degree (degreeFilter: tot/in/out; entityFilter: node_id, label, rel_id, rel_type)
Asymmetric/Symmetric Reachability (directed vs undirected for all or specific relTypes)
others to implement
:params{
  "cypher": {
    "regex" : ".*h.*",
    "cypher2" : "toedit"
     },
  "subgraph": {
     "startNode": (<id>|Node|list), 
     "maxLevel",
      relationshipFilter,
      labelFilter,
      bfs:true, 
      filterStartNode:false, 
      limit:-1, 
      endNodes:[], 
      terminatorNodes:[], 
      sequence, 
      beginSequenceAtStart:true
      },
  "degree": {
      "mode" : ("in"/"out"/"tot")*, //*min, max, pos, neg
      "maxLevel",
      "relationshipFilter",
      labelFilter,
      bfs:true, 
Query: toedit

//apoc subgraph
CALL apoc.path.subgraphAll(
    startNode <id>|Node|list, 
    {
        maxLevel: $ subgraph.maxLevel,
        relationshipFilter: $ subgraph.relationshipFilter,
        labelFilter: $ subgraph.labelFilter,
        bfs: $ subgraph.bfs, 
        filterStartNode: $ subgraph.filterStartNode, 
        limit: $ subgraph.limit, 
        endNodes: $ subgraph.endNodes, 
        terminatorNodes: $ subgraph.terminatorNodes, 
        sequence: $ subgraph.sequence, 
        beginSequenceAtStart: $ subgraph.beginSequenceAtStart
     }
) yield nodes, relationships

//degree
MATCH (n)
WHERE  $ degLabel IN LABELS(n)
RETURN apoc.node.degree(n, "FOLLOWS>") AS output;
//Subqueries
CALL {
apoc.node.degree.in(node, relationshipName)
apoc.node.degree.out(node, relationshipName)
  //degreeCF
  WITH nodes,relationships
  UNWIND nodes AS n
  WITH nodes,relationships,n
  MATCH (n: {name: "Michael"})
  WHERE  $ degLabel IN LABELS(n) AND 
  RETURN apoc.node.degree(p, "FOLLOWS>") AS output;
  WHERE degree(n)<=> $ degree
UNION
  MATCH (p:Person)
  RETURN p
  ORDER BY p.age DESC
  LIMIT 1
}
RETURN p.name, p.age
ORDER BY p.name

//relTypeCF
WITH nodes,relationships
UNWIND nodes AS n
WITH nodes,relationships,n
MATCH(n)-[r]-[m]
//nodeLabelCF
WITH nodes,relationships
UNWIND nodes AS n
WITH nodes,relationships,n
WHERE  $ label1 IN LABELS(n)
RETURN

Policy Flows:
This allows for policies to be defined in a configurable way that we already have implemented code-wise for authflows. Also, we should be able to reduce the overlapping resource set definition since we set some schema constraints and the visual process can prevent bad policies to be defined. Still part of the problem remains, as well as TOEDIT
Each flow block contains:
blueprint: ref_schema ?if unique no need of having this TOEDIT
op: apoc_subgraph(params),
conditions: [conds]
Flow Operations
Sets operators: union, intersection, difference (no complementarity).
Block Operations
WHERE clause filter on nodes:
Params:
in_nodes: input nodes
regex: 
list and pattern comprehension: 
UNWIND  $ in_nodes AS n
MATCH(n)
WHEREdt
RETURN [] 
subgraph filter
CALL apoc.path.subgraphAll(
    dt, 
    {
        maxLevel: 1 // $ maxLevel,
        relationshipFilter: $ relationshipFilter,
        labelFilter: $ labelFilter
    })
YIELD nodes,relationships
node degree
others

Conclusions
We are testing how a single cypher may behave in policy verification when the policy data is partially given by:
an edited schema of the KG, called Blueprint toedit;
policy params coming from policy data.
Also, the KG data may vary in such a way that the cypher does not capture anymore the correct resource set.
This lead us to think of the problem at least one in another, the inner being the Resource Set search within the KG, that is changing and for which we just assume to have the correct starting anchor (and node labels, rel types and optional depth range, plus some other to implement, but might not capture the RS change).
The first subtask to solve correctly is then the RS change, then we can think about the schema change (started addressing it with the Blueprint).
Next
In a subsequent task this can continue by:
expanding the FXC KG Schema, Data and Policies to highlight fails in the current policy verification query
look at solution
implement in FXC KG
build up starting REMA KG to start verifying policies and query in other KG domain
try the big filter cypher approach if the subgraph based query fails
maybe implement a solution mimicking policy flows if also a big filter cypher fails in the latest tests, considering now a fixed flow represented by a unique cypher with subqueries? It would be like a specific big filter cypher, with an engineered flow of subqueries. The real policy flow would be drawn and edited by the customer so it could change and it does change from policy to policy obviously. What cons does it bring along with the pros?
wiki this (and collaterals) back

Notes
From Conclusions and Next in the APOC cypher paragraph

update Miro back
Ontologies
Compliance: kgraph schema, is the business data schema; the blueprint is the policy data schema.
NOTE: eIDAS might become another compliance ontology sitting alongside the KG
GDS:
In general GDS pipelines which from an input graph project, transform and integrate data back into it, relaying the tasks logic to the pipelines.
entity resolution
signed graph projection for correlation clustering
weak and strong conn components
others using undirected projections
AI:
First approach (from ref1 and ref2):
ml models trained on a first period from the whole customer and then fine-tuned to data of an individual sub-space or individual users to understand the policy
the focus of the model shifts from fraud detection to identity/access assurance. In other  words, how confident can you be that a given user is who he or she  claims to be, based on past successful auths/authZs? This confidence  is gained by learning from data in users’ past interactions, such as:
Location/network
Time of day
Device fingerprint
Pattern of access
Keystroke dynamics
Much of this data is the same as what’s used to establish static  business context rules for access requests. The difference lies in how  the data is applied. Instead of declaring a network address “trusted,”  for example, the idea is to analyze user activity to determine if the  user has repeatedly provided a high level of authentication from a  particular IP address. That provides one data point to consider when  determining confidence in the user’s identity. Combine it with other  data points, creating confidence across multiple attributes, and you can  make an informed determination about whether more authentication is  needed.
Second approach (from ref3):
identity access to authorized resources may be seen of a very strict case of Recommendation  No, seriously!
Third approach (from ref1 and ref2):
all the authorization entities are topologically expanded with their function(s) and attributes embedded in the topology. ID triangles for example can be expanded and used as private key of a scheme similar to functional cryptography.
Katana might be tested to recognize the entities and also correlate them univocally (explain katana flow)

:link: Linked tasks: #202, #208,

**Comment by adelefelicia** [comment]
**Commented by: @indyboox Jul 01 2022 at 11:05 CET**

@pepe-zamarripa @bayegan I would like to continue this by:

- expanding the FXC KG Schema, Data and Policies to highlight fails in the current policy verification query
- look at solution
- implement in FXC KG
- build up starting REMA KG to start verifying policies and query in other KG domain
- maybe implement a solution mimicking policy flows



I would propose to then open a Part3, addressing these. Thank you!
